File: src/gobby/code_index/parser.py
"""Tree-sitter AST parser for code symbol extraction.

Parses source files into Symbol, ImportRelation, and CallRelation objects
using tree-sitter queries defined in the language registry.
"""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

try:
    from tree_sitter import Query, QueryCursor
except ImportError:
    Query = None  # type: ignore[assignment,misc]
    QueryCursor = None  # type: ignore[assignment,misc]

from gobby.code_index.hasher import file_content_hash, symbol_content_hash
from gobby.code_index.languages import (
    LANGUAGE_SPECS,
    detect_language,
    get_language_obj,
    get_parser_for_language,
)
from gobby.code_index.models import (
    CallRelation,
    ImportRelation,
    ParseResult,
    Symbol,
)
from gobby.code_index.security import (
    has_secret_extension,
    is_binary,
    is_symlink_safe,
    should_exclude,
    validate_path,
)
from gobby.config.code_index import CodeIndexConfig

logger = logging.getLogger(__name__)


class CodeParser:
    """Parses source files into symbols using tree-sitter."""

    def __init__(self, config: CodeIndexConfig) -> None:
        if Query is None:
            raise ImportError(
                "tree-sitter is required for code indexing. "
                "Install it with: uv pip install tree-sitter-language-pack"
            )
        self.config = config
        self._supported_languages = set(config.languages)

    def parse_file(
        self,
        file_path: str,
        project_id: str,
        root_path: str,
    ) -> ParseResult | None:
        """Parse a single file into symbols, imports, and calls.

        Returns None if file should be skipped (binary, excluded, etc).
        """
        path = Path(file_path)
        root = Path(root_path)

        # Security checks
        if not validate_path(path, root):
            return None
        if not is_symlink_safe(path, root):
            return None
        if should_exclude(path, self.config.exclude_patterns):
            return None
        if has_secret_extension(path):
            return None

        # Size check
        try:
            size = path.stat().st_size
            if size > self.config.max_file_size_bytes or size == 0:
                return None
        except OSError:
            return None

        # Binary check
        if is_binary(path):
            return None

        # Language detection
        language = detect_language(file_path)
        if language is None or language not in self._supported_languages:
            return None

        spec = LANGUAGE_SPECS.get(language)
        if spec is None:
            return None

        # Get parser
        parser = get_parser_for_language(language)
        if parser is None:
            return None

        # Read and parse
        try:
            source = path.read_bytes()
        except OSError as e:
            logger.debug(f"Cannot read {file_path}: {e}")
            return None

        try:
            tree = parser.parse(source)
        except Exception as e:
            logger.warning(f"Parse error for {file_path}: {e}")
            return None

        # Make path relative to root for storage
        try:
            rel_path = str(path.resolve().relative_to(root.resolve()))
        except ValueError:
            rel_path = str(path)

        # Extract symbols
        symbols = self._extract_symbols(
            tree, source, spec, language, project_id, rel_path
        )

        # Extract imports
        imports = self._extract_imports(tree, source, spec, rel_path)

        # Extract calls
        calls = self._extract_calls(tree, source, spec, language, rel_path, symbols)

        return ParseResult(symbols=symbols, imports=imports, calls=calls)

    def _extract_symbols(
        self,
        tree: Any,
        source: bytes,
        spec: object,
        language: str,
        project_id: str,
        rel_path: str,
    ) -> list[Symbol]:
        """Extract symbols from parsed tree using language-specific queries."""
        if not hasattr(spec, "symbol_query") or not spec.symbol_query.strip():
            return []

        lang_obj = get_language_obj(language)
        if lang_obj is None:
            return []

        try:
            query = Query(lang_obj, spec.symbol_query)
        except Exception as e:
            logger.debug(f"Query compilation failed for {language}: {e}")
            return []

        root_node = tree.root_node
        cursor = QueryCursor(query)
        matches = cursor.matches(root_node)

        symbols: list[Symbol] = []
        seen_ids: set[str] = set()

        for _, captures in matches:
            # Find the name capture and the definition capture
            name_node = None
            def_node = None
            kind = "function"

            for cap_name, nodes in captures.items():
                node = nodes[0] if isinstance(nodes, list) else nodes
                if cap_name == "name":
                    name_node = node
                elif cap_name.startswith("definition."):
                    def_node = node
                    kind = cap_name.split(".", 1)[1]

            if name_node is None or def_node is None:
                continue

            name = source[name_node.start_byte:name_node.end_byte].decode(
                "utf-8", errors="replace"
            )

            # Build qualified name (will be updated for nested symbols below)
            qualified_name = name

            # Extract signature (first line of the definition)
            start_line_bytes = source[
                def_node.start_byte:
                source.find(b"\n", def_node.start_byte)
                if source.find(b"\n", def_node.start_byte) != -1
                else def_node.end_byte
            ]
            signature = start_line_bytes.decode("utf-8", errors="replace").strip()
            if len(signature) > 200:
                signature = signature[:200] + "..."

            # Extract docstring (look for string as first child of body)
            docstring = self._extract_docstring(def_node, source, language)

            # Compute content hash
            c_hash = symbol_content_hash(source, def_node.start_byte, def_node.end_byte)

            symbol_id = Symbol.make_id(
                project_id, rel_path, name, kind, def_node.start_byte
            )

            if symbol_id in seen_ids:
                continue
            seen_ids.add(symbol_id)

            symbols.append(
                Symbol(
                    id=symbol_id,
                    project_id=project_id,
                    file_path=rel_path,
                    name=name,
                    qualified_name=qualified_name,
                    kind=kind,
                    language=language,
                    byte_start=def_node.start_byte,
                    byte_end=def_node.end_byte,
                    line_start=def_node.start_point[0] + 1,
                    line_end=def_node.end_point[0] + 1,
                    signature=signature,
                    docstring=docstring,
                    content_hash=c_hash,
                )
            )

        # Link parent symbols (methods inside classes)
        self._link_parents(symbols, spec)

        return symbols

    def _link_parents(self, symbols: list[Symbol], spec: object) -> None:
        """Set parent_symbol_id and qualified_name for nested symbols."""
        # Sort by byte_start so parents come before children
        sorted_syms = sorted(symbols, key=lambda s: s.byte_start)

        # Build a stack of potential parents
        for i, sym in enumerate(sorted_syms):
            # Look for the nearest enclosing class/container
            for j in range(i - 1, -1, -1):
                parent = sorted_syms[j]
                if parent.kind in ("class", "type") and (
                    parent.byte_start <= sym.byte_start
                    and parent.byte_end >= sym.byte_end
                ):
                    sym.parent_symbol_id = parent.id
                    sym.qualified_name = f"{parent.name}.{sym.name}"
                    if sym.kind == "function":
                        sym.kind = "method"
                    break

    def _extract_docstring(
        self, node: Any, source: bytes, language: str
    ) -> str | None:
        """Try to extract a docstring from a definition node."""
        if language not in ("python", "javascript", "typescript"):
            return None

        # For Python: look for expression_statement > string as first child of body
        body = None
        for child in node.children:
            if hasattr(child, "type") and child.type in ("block", "statement_block"):
                body = child
                break

        if body is None:
            return None

        for child in body.children:
            if not hasattr(child, "type"):
                continue
            # Skip comments, newlines
            if child.type in ("comment", "\n", "newline"):
                continue

            # Find the string node — may be direct child or inside expression_statement
            string_node = None
            if child.type == "string":
                string_node = child
            elif child.type == "expression_statement":
                for grandchild in child.children:
                    if hasattr(grandchild, "type") and grandchild.type == "string":
                        string_node = grandchild
                        break

            if string_node is None:
                break

            # Extract text — try string_content child first (newer tree-sitter),
            # fall back to stripping quotes from full text
            for sc in string_node.children:
                if hasattr(sc, "type") and sc.type == "string_content":
                    raw = source[sc.start_byte:sc.end_byte].decode(
                        "utf-8", errors="replace"
                    )
                    return raw.strip() if raw.strip() else None

            raw = source[string_node.start_byte:string_node.end_byte].decode(
                "utf-8", errors="replace"
            )
            for q in ('"""', "'''", '"', "'"):
                if raw.startswith(q) and raw.endswith(q):
                    raw = raw[len(q):-len(q)]
                    break
            return raw.strip() if raw.strip() else None

        return None

    def _extract_imports(
        self,
        tree: Any,
        source: bytes,
        spec: object,
        rel_path: str,
    ) -> list[ImportRelation]:
        """Extract import statements."""
        if not hasattr(spec, "import_query") or not spec.import_query.strip():
            return []

        lang_name = None
        for lname, lspec in LANGUAGE_SPECS.items():
            if lspec is spec:
                lang_name = lname
                break
        if lang_name is None:
            return []

        lang_obj = get_language_obj(lang_name)
        if lang_obj is None:
            return []

        try:
            query = Query(lang_obj, spec.import_query)
        except Exception:
            return []

        cursor = QueryCursor(query)
        matches = cursor.matches(tree.root_node)
        imports: list[ImportRelation] = []

        for _, captures in matches:
            for _cap_name, nodes in captures.items():
                node = nodes[0] if isinstance(nodes, list) else nodes
                text = source[node.start_byte:node.end_byte].decode(
                    "utf-8", errors="replace"
                )
                # Simple extraction: store the full import text as target_module
                imports.append(
                    ImportRelation(
                        source_file=rel_path,
                        target_module=text.strip(),
                        imported_names=[],
                    )
                )

        return imports

    def _extract_calls(
        self,
        tree: Any,
        source: bytes,
        spec: Any,
        language: str,
        rel_path: str,
        symbols: list[Symbol],
    ) -> list[CallRelation]:
        """Extract function/method calls."""
        if not hasattr(spec, "call_query") or not spec.call_query.strip():
            return []

        lang_obj = get_language_obj(language)
        if lang_obj is None:
            return []

        try:
            query = Query(lang_obj, spec.call_query)
        except Exception:
            return []

        cursor = QueryCursor(query)
        matches = cursor.matches(tree.root_node)
        calls: list[CallRelation] = []

        for _, captures in matches:
            name_node = None
            call_node = None
            for cap_name, nodes in captures.items():
                node = nodes[0] if isinstance(nodes, list) else nodes
                if cap_name == "name":
                    name_node = node
                elif cap_name == "call":
                    call_node = node

            if name_node is None:
                continue

            callee_name = source[name_node.start_byte:name_node.end_byte].decode(
                "utf-8", errors="replace"
            )

            # Find enclosing symbol (caller)
            target_node = call_node or name_node
            caller_id = ""
            for sym in symbols:
                if sym.byte_start <= target_node.start_byte <= sym.byte_end:
                    caller_id = sym.id

            calls.append(
                CallRelation(
                    caller_symbol_id=caller_id,
                    callee_name=callee_name,
                    file_path=rel_path,
                    line=name_node.start_point[0] + 1,
                )
            )

        return calls

    def get_file_hash(self, file_path: str) -> str | None:
        """Get content hash for a file, or None if unreadable."""
        try:
            return file_content_hash(file_path)
        except OSError:
            return None

File: rust/gcode/src/index/parser.rs
//! Tree-sitter AST parsing for symbol, import, and call extraction.
//! Ports logic from src/gobby/code_index/parser.py.

use std::collections::HashSet;
use std::path::Path;

use streaming_iterator::StreamingIterator;
use tree_sitter::{Parser, Query, QueryCursor};

use crate::index::hasher::symbol_content_hash;
use crate::index::languages;
use crate::index::security;
use crate::models::{CallRelation, ImportRelation, ParseResult, Symbol};

/// Maximum file size to index (10 MB).
const MAX_FILE_SIZE: u64 = 10 * 1024 * 1024;

/// Parse a single file into symbols, imports, and calls.
/// Returns None if the file should be skipped.
pub fn parse_file(
    file_path: &Path,
    project_id: &str,
    root_path: &Path,
    exclude_patterns: &[String],
) -> Option<ParseResult> {
    // Security checks
    if !security::validate_path(file_path, root_path) {
        return None;
    }
    if !security::is_symlink_safe(file_path, root_path) {
        return None;
    }
    if security::should_exclude(file_path, exclude_patterns) {
        return None;
    }
    if security::has_secret_extension(file_path) {
        return None;
    }

    let meta = file_path.metadata().ok()?;
    if meta.len() == 0 || meta.len() > MAX_FILE_SIZE {
        return None;
    }

    if security::is_binary(file_path) {
        return None;
    }

    let file_str = file_path.to_string_lossy();
    let language = languages::detect_language(&file_str)?;
    let spec = languages::get_spec(language)?;
    let ts_lang = languages::get_ts_language(language)?;

    let source = std::fs::read(file_path).ok()?;

    let mut parser = Parser::new();
    parser.set_language(&ts_lang).ok()?;
    let tree = parser.parse(&source, None)?;

    let rel_path = file_path
        .canonicalize()
        .ok()
        .and_then(|abs| {
            root_path
                .canonicalize()
                .ok()
                .and_then(|root| {
                    abs.strip_prefix(&root)
                        .ok()
                        .map(|p| p.to_string_lossy().to_string())
                })
        })
        .unwrap_or_else(|| file_str.to_string());

    let mut symbols =
        extract_symbols(&tree, &source, spec, language, &ts_lang, project_id, &rel_path);
    link_parents(&mut symbols);
    let imports = extract_imports(&tree, &source, spec, &ts_lang, &rel_path);
    let calls = extract_calls(&tree, &source, spec, &ts_lang, &rel_path, &symbols);

    Some(ParseResult {
        symbols,
        imports,
        calls,
    })
}

fn extract_symbols(
    tree: &tree_sitter::Tree,
    source: &[u8],
    spec: &languages::LanguageSpec,
    language: &str,
    ts_lang: &tree_sitter::Language,
    project_id: &str,
    rel_path: &str,
) -> Vec<Symbol> {
    if spec.symbol_query.trim().is_empty() {
        return Vec::new();
    }

    let query = match Query::new(ts_lang, spec.symbol_query) {
        Ok(q) => q,
        Err(_) => return Vec::new(),
    };

    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree.root_node(), source);

    let mut symbols = Vec::new();
    let mut seen_ids = HashSet::new();
    let capture_names: Vec<String> = query.capture_names().iter().map(|s| s.to_string()).collect();

    while let Some(m) = matches.next() {
        let mut name_text: Option<String> = None;
        let mut def_node = None;
        let mut kind = String::from("function");

        for cap in m.captures {
            let cap_name = &capture_names[cap.index as usize];
            if cap_name == "name" {
                name_text = Some(
                    String::from_utf8_lossy(&source[cap.node.start_byte()..cap.node.end_byte()])
                        .to_string(),
                );
            } else if let Some(k) = cap_name.strip_prefix("definition.") {
                def_node = Some(cap.node);
                kind = k.to_string();
            }
        }

        let (name, node) = match (name_text, def_node) {
            (Some(n), Some(d)) => (n, d),
            _ => continue,
        };

        // Signature: first line of definition
        let sig_end = source[node.start_byte()..]
            .iter()
            .position(|&b| b == b'\n')
            .map(|p| node.start_byte() + p)
            .unwrap_or(node.end_byte());
        let mut signature =
            String::from_utf8_lossy(&source[node.start_byte()..sig_end])
                .trim()
                .to_string();
        if signature.len() > 200 {
            signature.truncate(200);
            signature.push_str("...");
        }

        let docstring = extract_docstring(&node, source, language);
        let c_hash = symbol_content_hash(source, node.start_byte(), node.end_byte());
        let symbol_id =
            Symbol::make_id(project_id, rel_path, &name, &kind, node.start_byte());

        if seen_ids.contains(&symbol_id) {
            continue;
        }
        seen_ids.insert(symbol_id.clone());

        symbols.push(Symbol {
            id: symbol_id,
            project_id: project_id.to_string(),
            file_path: rel_path.to_string(),
            name: name.clone(),
            qualified_name: name,
            kind,
            language: language.to_string(),
            byte_start: node.start_byte(),
            byte_end: node.end_byte(),
            line_start: node.start_position().row + 1,
            line_end: node.end_position().row + 1,
            signature: Some(signature),
            docstring,
            parent_symbol_id: None,
            content_hash: c_hash,
            summary: None,
            created_at: String::new(),
            updated_at: String::new(),
        });
    }

    symbols
}

fn link_parents(symbols: &mut [Symbol]) {
    let mut indices: Vec<usize> = (0..symbols.len()).collect();
    indices.sort_by_key(|&i| symbols[i].byte_start);

    for idx in 0..indices.len() {
        let i = indices[idx];
        for jdx in (0..idx).rev() {
            let j = indices[jdx];
            let parent_kind = symbols[j].kind.as_str();
            if (parent_kind == "class" || parent_kind == "type")
                && symbols[j].byte_start <= symbols[i].byte_start
                && symbols[j].byte_end >= symbols[i].byte_end
            {
                let parent_name = symbols[j].name.clone();
                let parent_id = symbols[j].id.clone();
                let sym = &mut symbols[i];
                sym.parent_symbol_id = Some(parent_id);
                sym.qualified_name = format!("{}.{}", parent_name, sym.name);
                if sym.kind == "function" {
                    sym.kind = "method".to_string();
                }
                break;
            }
        }
    }
}

fn extract_docstring(node: &tree_sitter::Node, source: &[u8], language: &str) -> Option<String> {
    if !matches!(language, "python" | "javascript" | "typescript") {
        return None;
    }

    let mut body = None;
    let mut walk = node.walk();
    for child in node.children(&mut walk) {
        let ty = child.kind();
        if ty == "block" || ty == "statement_block" {
            body = Some(child);
            break;
        }
    }
    let body = body?;

    let mut walk2 = body.walk();
    for child in body.children(&mut walk2) {
        let ty = child.kind();
        if ty == "comment" || ty == "\n" || ty == "newline" {
            continue;
        }

        let string_node = if ty == "string" {
            Some(child)
        } else if ty == "expression_statement" {
            let mut w3 = child.walk();
            child.children(&mut w3).find(|gc| gc.kind() == "string")
        } else {
            None
        };

        let string_node = match string_node {
            Some(n) => n,
            None => return None,
        };

        // Try string_content child first
        let mut w4 = string_node.walk();
        for sc in string_node.children(&mut w4) {
            if sc.kind() == "string_content" {
                let raw =
                    String::from_utf8_lossy(&source[sc.start_byte()..sc.end_byte()]);
                let trimmed = raw.trim();
                return if trimmed.is_empty() {
                    None
                } else {
                    Some(trimmed.to_string())
                };
            }
        }

        // Fallback: strip quotes
        let raw = String::from_utf8_lossy(
            &source[string_node.start_byte()..string_node.end_byte()],
        );
        let raw = raw.trim();
        let stripped = strip_quotes(raw);
        return if stripped.is_empty() {
            None
        } else {
            Some(stripped.to_string())
        };
    }

    None
}

fn strip_quotes(s: &str) -> &str {
    for q in &["\"\"\"", "'''", "\"", "'"] {
        if s.starts_with(q) && s.ends_with(q) && s.len() >= q.len() * 2 {
            return s[q.len()..s.len() - q.len()].trim();
        }
    }
    s
}

fn extract_imports(
    tree: &tree_sitter::Tree,
    source: &[u8],
    spec: &languages::LanguageSpec,
    ts_lang: &tree_sitter::Language,
    rel_path: &str,
) -> Vec<ImportRelation> {
    if spec.import_query.trim().is_empty() {
        return Vec::new();
    }

    let query = match Query::new(ts_lang, spec.import_query) {
        Ok(q) => q,
        Err(_) => return Vec::new(),
    };

    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree.root_node(), source);
    let capture_names: Vec<String> = query.capture_names().iter().map(|s| s.to_string()).collect();
    let mut imports = Vec::new();

    while let Some(m) = matches.next() {
        for cap in m.captures {
            let cap_name = &capture_names[cap.index as usize];
            if cap_name == "import" {
                let text = String::from_utf8_lossy(
                    &source[cap.node.start_byte()..cap.node.end_byte()],
                )
                .trim()
                .to_string();
                imports.push(ImportRelation {
                    file_path: rel_path.to_string(),
                    module_name: text,
                    project_id: String::new(),
                });
            }
        }
    }

    imports
}

fn extract_calls(
    tree: &tree_sitter::Tree,
    source: &[u8],
    spec: &languages::LanguageSpec,
    ts_lang: &tree_sitter::Language,
    rel_path: &str,
    symbols: &[Symbol],
) -> Vec<CallRelation> {
    if spec.call_query.trim().is_empty() {
        return Vec::new();
    }

    let query = match Query::new(ts_lang, spec.call_query) {
        Ok(q) => q,
        Err(_) => return Vec::new(),
    };

    let mut cursor = QueryCursor::new();
    let mut matches = cursor.matches(&query, tree.root_node(), source);
    let capture_names: Vec<String> = query.capture_names().iter().map(|s| s.to_string()).collect();
    let mut calls = Vec::new();

    while let Some(m) = matches.next() {
        let mut name_node = None;
        let mut call_node = None;

        for cap in m.captures {
            let cap_name = &capture_names[cap.index as usize];
            if cap_name == "name" {
                name_node = Some(cap.node);
            } else if cap_name == "call" {
                call_node = Some(cap.node);
            }
        }

        let name_n = match name_node {
            Some(n) => n,
            None => continue,
        };

        let callee_name =
            String::from_utf8_lossy(&source[name_n.start_byte()..name_n.end_byte()]).to_string();

        let target = call_node.unwrap_or(name_n);
        let caller_id = symbols
            .iter()
            .filter(|s| {
                s.byte_start <= target.start_byte() && target.start_byte() <= s.byte_end
            })
            .last()
            .map(|s| s.id.clone())
            .unwrap_or_default();

        calls.push(CallRelation {
            caller_id,
            callee_name,
            file_path: rel_path.to_string(),
            line: name_n.start_position().row + 1,
            project_id: String::new(),
        });
    }

    calls
}
